controller/orphan_controller.go

```diff
@@ -269,48 +303,187 @@ func (oc *OrphanController) isResponsibleFor(orphan *longhorn.Orphan) bool {
+	defer func() {
+		if err != nil {
+			err = errors.Wrapf(err, "cannot cleanup replica instance orphan %v", orphan.Name)
+		}
+	}()
+
+	instance, deType, err := oc.extractOrphanedRuntimeInfo(orphan)
+	if err != nil {
+		return false, err
+	}
+
+	var spec *longhorn.InstanceSpec
+	if replicaCR, err := oc.ds.GetReplicaRO(instance); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return false, err
+		}
+		spec = nil
+	} else {
+		spec = &replicaCR.Spec.InstanceSpec
+	}
+	return oc.cleanupOrphanedRuntimeInstance(orphan, instance, longhorn.InstanceManagerTypeReplica, deType, spec)
+}
+
+func (oc *OrphanController) cleanupOrphanedRuntimeInstance(orphan *longhorn.Orphan, instance string, imType longhorn.InstanceManagerType, deType longhorn.DataEngineType, instanceCRSpec *longhorn.InstanceSpec) (isCleanupComplete bool, err error) {
+	if instanceCRSpec != nil && instanceCRSpec.NodeID == oc.controllerID {
+		// Instance is scheduled back to current node and is no longer considered orphaned.
+		// Finalize the orphan CR directly without terminating the instance process.
+		return true, nil
+	}
+
+	im, err := oc.ds.GetRunningInstanceManagerByNodeRO(oc.controllerID, deType)
+	if err != nil {
+		oc.logger.WithError(err).Infof("No running instance manager for node %v for deleting runtime instance orphan %v", oc.controllerID, orphan.Name)
+		return true, nil
+	}
+	imc, err := engineapi.NewInstanceManagerClient(im, false)
+	if err != nil {
+		return false, err
+	}
+	defer func() {
+		if closeErr := imc.Close(); closeErr != nil {
+			oc.logger.WithError(closeErr).Error("failed to close instance manager client")
+		}
+	}()
+
+	runtimeInstanceExist := types.GetCondition(orphan.Status.Conditions, longhorn.OrphanConditionTypeRuntimeInstanceState).Status == longhorn.ConditionStatusTrue
+	if runtimeInstanceExist {
+		if err = oc.deleteRuntimeInstance(imc, instance, imType, deType); err != nil {
+			return false, err
+		}
+	}
+	return oc.confirmOrphanRuntimeInstanceCleanup(imc, instance, imType, deType)
+}
+
+func (oc *OrphanController) extractOrphanedRuntimeInfo(orphan *longhorn.Orphan) (name string, dataEngineType longhorn.DataEngineType, err error) {
+	name, ok := orphan.Spec.Parameters[longhorn.OrphanInstanceName]
+	if !ok {
+		return "", "", fmt.Errorf("failed to get instance name for runtime instance orphan %v", orphan.Name)
+	}
+
+	dataEngineTypeStr, ok := orphan.Spec.Parameters[longhorn.OrphanDataEngineType]
+	if !ok {
+		return "", "", fmt.Errorf("failed to get data engine type for runtime instance orphan %v", orphan.Name)
+	}
+	switch dataEngineType = longhorn.DataEngineType(dataEngineTypeStr); dataEngineType {
+	case longhorn.DataEngineTypeV1, longhorn.DataEngineTypeV2:
+		// supported data engine type
 	default:
-		err = fmt.Errorf("unknown orphan type %v", orphan.Spec.Type)
+		return "", "", fmt.Errorf("unknown data engine type %v for runtime instance orphan %v", dataEngineTypeStr, orphan.Name)
```

Comment:
Is it possible to validate via webhook?

Reply:
Good idea, will add this.

Reply:
Where is the validation of `dataEngineType` in webhook?

Reply:
Sorry for misleading. We have checked orphan type in webhook, but not the data engine type.

Though the type is checked in the webhook, printing error log for unknown type here would prevent developers forgot check this in the webhook in the future. At least we'll get errors, instead of leaving it silently.

Reply:
Just add checks for `dataEngineType`. Thanks for reminding.